core/src/identity.rs

```diff
@@ -115,6 +115,37 @@ impl Keypair {
             Secp256k1(pair) => PublicKey::Secp256k1(pair.public().clone()),
         }
     }
+
+    /// Decode a private key from a protobuf structure and parse it as a [`Keypair`].
+    pub fn from_protobuf_encoding(bytes: &[u8]) -> Result<Keypair, DecodingError> {
+        use prost::Message;
+
+        let mut private_key = keys_proto::PrivateKey::decode(bytes)
+            .map_err(|e| DecodingError::new("Protobuf").source(e))
+            .map(zeroize::Zeroizing::new)?;
+
+        let key_type = keys_proto::KeyType::from_i32(private_key.r#type)
+            .ok_or_else(|| DecodingError::new(format!("unknown key type: {}", private_key.r#type)))?;
+
+        match key_type {
+            keys_proto::KeyType::Ed25519 => {
+                ed25519::Keypair::decode(&mut private_key.data).map(Keypair::Ed25519)
+            },
+            keys_proto::KeyType::Rsa => {
+                Err(DecodingError::new("Unsupported"))
```

Comment:
Would it make sense to add a more elaborate error message like "RSA keys are unsupported"? (same below).

Reply:
Good idea. Added via https://github.com/libp2p/rust-libp2p/pull/2090/commits/42ff89ebc5402dd8a8aaa6f7a7dc9ef237c3491a.